### ISSUE TITLE
Much higher CPU Usage on Linux than on Windows

### ISSUE THREAD

**Issue by BrunoJuchli** [dotnet/runtime#71921]
### Description

We've just moved some of our software to run on Linux instead of Windows nodes and the CPU usage increased significantly.

For our testing system, which is mostly idling:
Average CPU load:
![image](https://user-images.githubusercontent.com/2090172/178202279-af531f43-3997-45a7-87a8-f49a88373108.png)
Windows: 17mc
Linux: 49-60mc (mostly the bar is at 58)

95th Percentile CPU load:
![image](https://user-images.githubusercontent.com/2090172/178201696-c9095463-22ba-48e2-93ff-a3887cc16d50.png)
Windows: 41mc
Linux: 226-231mc
Increase of 5.5x

And also in production we can see a similar behavior (I'm not including graphs here because I had to increase the CPU limit for Linux nodes, so the Azure Insights graphs would give a wrong impression):

- Average CPU Load:
  - Windows: 310 mCPU
  - Linux: 1201 mCPU
  - increase of 3.9x

I've sanity checked the numbers given by Azure Insights using `dotnet-counters` which shows correspondingly high cpu usage (%).

We've got other software which we've had running on .net 6 Linux for a longer time now where we didn't notice this issue.
The affected software, 4 different applications, all follow the same pattern. There's a number of items to be processed periodically and the load is distributed across multiple instances of the application.
The basic implementation is a `while(true)` loop checking for work; at the end of each iteration there is a `Thread.Sleep(..)` with a sleep time between 200ms and 1s (I also changed this to an async implementation with `Task.Delay` without any obvious effect).
We're also using Redis (StackExchange.Redis) for Pub Sub. Because of bad design we're unnecessarily creating/running 49 `RedisChannel`s, which, at least in the production system, creates some unnecessary load - but does not explain the difference between Windows and Linux. 
Also, for each of the 49 `RedisChannel` instances we create a publish queue (`BlockingCollection`) and a Task which publishes the queued items on the `RedisChannel`. The task is created via `new Task(Action).Start()`.
That's not the end of it, but I don't know all the wheels in this machine myself yet.

I searched several hours on the .net for known performance deterioration of .net /5/6 on Linux but couldn't find anything which seemed possibly relevant and actionable (i.E. change something and observe the effects). 

### Configuration
Build: Release config built on SDK 6.0.301 (on a github Windows Server 2022 machine for deployments built for Windows Nodes and a github Linux machine for stuff deployed to Linux Nodes).

Runtime:

Run via Kubernetes 1.23.5 on Azure.
For testing: B2ms (Windows) and B2s (Linux) nodes respectively.
Production: F8s_v2 nodes for both and Windows and Linux.

#### Windows
dotnet 6.0.6, Windows Server Core 2019.
Docker:

    FROM mcr.microsoft.com/dotnet/runtime:6.0-windowsservercore-ltsc2019
    ...
    ENTRYPOINT ["dotnet", "foo.dll"]

#### Linux
dotnet 6.0.6 on Alpine 3.16 amd64.
Docker:

    FROM mcr.microsoft.com/dotnet/runtime:6.0-alpine
    RUN apk update && \
      apk add --no-cache icu tzdata;
    ENV \
      DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=false \
      LC_ALL=en_US.UTF-8 \
      LANG=en_US.UTF-8
    ENTRYPOINT ["dotnet", "foo.dll"]

### Analysis

Possibly related to https://github.com/dotnet/runtime/issues/43063

**Comment by EgorBo** [comment]
You attached two traces but both of them seem to be from Linux, was one of them supposed to be windows to compare?

**Comment by BrunoJuchli** [comment]
@EgorBo 
No, they are both from Linux. One's a bit shorter and the other longer.
I've now also added a trace of windows; though I think I took this from my local machine. I'll retake that one from a test system so that they are better comparable.

**Comment by BrunoJuchli** [comment]
@EgorBo 
I added traces and counters for linux and windows in comparable idle situation.

**Comment by BrunoJuchli** [comment]
@mangod9 
I've looked a bit closer at it and had to correct my last comment. At the moment I don't have a clear indication that would point to StackExchange.Redis - it's being used in affected and unaffected apps.

**Comment by kouvel** [comment]
> Is my understanding correct that the traces generated by dotnet-trace don't directly relate to actual cpu time consumed/cpu-load consumed but rather just show the "real" time a thread spends in a given method? i.E. it's quite difficult to find out what's causing actual cpu load?

Yes, `dotnet-trace` samples managed thread stacks and it cannot know whether a thread is blocked or not, so the sample count will include blocked time, which can make it more difficult to analyze on-CPU usage. [`perfcollect`](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/trace-perfcollect-lttng) can produce on-CPU traces that may be easier to analyze for CPU usage.

I'm not able to open the Linux traces in the latest PerfView, I'm seeing an error:
```
Error: Exception EventPipe conversion: System.Exception: Read past end of stream.
[Error: exception while opening EventPipe data.]
```

If I try to ignore the errors with `-ContinueOnError`, I'm able to see some events but I'm not able to see sampled stacks. Are you able to open those traces in PerfView?

I see events indicating thread pool worker thread starvation in both the Windows and Linux traces. Example from the bulkmetersim Windows trace below, there are a few more in the Linux trace:
```
Event Name                                                                 	 Time MSec	Process Name          	Rest  
Microsoft-Windows-DotNETRuntime/ThreadPoolWorkerThreadAdjustment/Adjustment	15,513.017	Process(14840) (14840)	ThreadID="13,424" ProcessorNumber="0" AverageThroughput="0.000" NewWorkerThreadCount="44" Reason="Starvation" ClrInstanceID="0" 
```

Some callee stacks that appear to be blocking thread pool worker threads:

```
Name                                                                                                                                                                                                                                              	Inc %	         Inc
 System.Private.CoreLib.il!System.Threading.PortableThreadPool+WorkerThread.WorkerThreadStart()                                                                                                                                                   	 77.2	   1,481,728
+ System.Private.CoreLib.il!ThreadPoolWorkQueue.Dispatch                                                                                                                                                                                          	 70.1	   1,346,848
|+ System.Private.CoreLib.il!Task.ExecuteEntryUnsafe                                                                                                                                                                                              	 70.1	   1,346,717
||+ System.Private.CoreLib.il!Task.ExecuteWithThreadLocal                                                                                                                                                                                         	 70.1	   1,346,717
|| + System.Private.CoreLib.il!ExecutionContext.RunFromThreadPoolDispatchLoop                                                                                                                                                                     	 70.1	   1,346,717
||  + System.Private.CoreLib.il!System.Threading.Tasks.Task+<>c.<.cctor>b__272_0(class System.Object)                                                                                                                                             	 70.1	   1,346,717
||   + System.Private.CoreLib.il!Task.InnerInvoke                                                                                                                                                                                                 	 70.1	   1,346,717
||    + Datastore.Repositories!Datastore.Repositories.PubQueue`2[System.__Canon,System.Guid].Run()                                                                                                                                                	 45.7	     876,932
||    |+ System.Collections.Concurrent.il!System.Collections.Concurrent.BlockingCollection`1[System.__Canon].Take()                                                                                                                               	 45.7	     876,932
||    | + System.Collections.Concurrent.il!System.Collections.Concurrent.BlockingCollection`1[System.__Canon].TryTakeWithNoTimeValidation(!0&,int32,value class System.Threading.CancellationToken,class System.Threading.CancellationTokenSource)	 45.7	     876,932
||    |  + System.Private.CoreLib.il!SemaphoreSlim.Wait                                                                                                                                                                                           	 45.7	     876,932
||    |   + System.Private.CoreLib.il!SemaphoreSlim.WaitUntilCountOrTimeout                                                                                                                                                                       	 45.7	     876,932
||    |    + System.Private.CoreLib.il!Monitor.Wait                                                                                                                                                                                               	 45.7	     876,932
||    + Datastore.Repositories!Datastore.Repositories.PubQueue`2[System.__Canon,System.Int64].Run()                                                                                                                                               	 14.7	     281,871
||    |+ System.Collections.Concurrent.il!System.Collections.Concurrent.BlockingCollection`1[System.__Canon].Take()                                                                                                                               	 14.7	     281,871
||    | + System.Collections.Concurrent.il!System.Collections.Concurrent.BlockingCollection`1[System.__Canon].TryTakeWithNoTimeValidation(!0&,int32,value class System.Threading.CancellationToken,class System.Threading.CancellationTokenSource)	 14.7	     281,871
||    |  + System.Private.CoreLib.il!SemaphoreSlim.Wait                                                                                                                                                                                           	 14.7	     281,871
||    |   + System.Private.CoreLib.il!SemaphoreSlim.WaitUntilCountOrTimeout                                                                                                                                                                       	 14.7	     281,871
||    |    + System.Private.CoreLib.il!Monitor.Wait                                                                                                                                                                                               	 14.7	     281,871
||    + Datastore.Repositories!Datastore.Repositories.PubQueue`2[System.__Canon,PubSub.Interface.SmartMeConnectPubSubMessage].Run()                                                                                                               	  6.5	     125,276
||    |+ System.Collections.Concurrent.il!System.Collections.Concurrent.BlockingCollection`1[System.__Canon].Take()                                                                                                                               	  6.5	     125,276
||    | + System.Collections.Concurrent.il!System.Collections.Concurrent.BlockingCollection`1[System.__Canon].TryTakeWithNoTimeValidation(!0&,int32,value class System.Threading.CancellationToken,class System.Threading.CancellationTokenSource)	  6.5	     125,276
||    |  + System.Private.CoreLib.il!SemaphoreSlim.Wait                                                                                                                                                                                           	  6.5	     125,276
||    |   + System.Private.CoreLib.il!SemaphoreSlim.WaitUntilCountOrTimeout                                                                                                                                                                       	  6.5	     125,276
||    |    + System.Private.CoreLib.il!Monitor.Wait                                                                                                                                                                                               	  6.5	     125,276
||    + Datastore.Repositories!Datastore.Repositories.PubQueue`2[System.Guid,System.Guid].Run()                                                                                                                                                   	  1.6	      31,319
||    |+ System.Collections.Concurrent.il!System.Collections.Concurrent.BlockingCollection`1[System.Guid].Take()                                                                                                                                  	  1.6	      31,319
||    | + System.Collections.Concurrent.il!System.Collections.Concurrent.BlockingCollection`1[System.Guid].TryTake(!0&,int32,value class System.Threading.CancellationToken)                                                                      	  1.6	      31,319
||    |  + System.Collections.Concurrent.il!System.Collections.Concurrent.BlockingCollection`1[System.Guid].TryTakeWithNoTimeValidation(!0&,int32,value class System.Threading.CancellationToken,class System.Threading.CancellationTokenSource)  	  1.6	      31,319
||    |   + System.Private.CoreLib.il!SemaphoreSlim.Wait                                                                                                                                                                                          	  1.6	      31,319
||    |    + System.Private.CoreLib.il!SemaphoreSlim.WaitUntilCountOrTimeout                                                                                                                                                                      	  1.6	      31,319
||    |     + System.Private.CoreLib.il!Monitor.Wait                                                                                                                                                                                              	  1.6	      31,319
```

Blocking thread pool threads often leads to issues. As you can see in the `Adjustment` event above indicating `Starvation`, the average throughput of queued work items goes to zero. Threads aren't injected very quickly in those cases. Although I see more frequent starvation in the Linux trace, the event only indicates starvation when all thread pool threads have not completed a work item in ~0.5-1 s. It doesn't indicate how many threads are blocked on average. How many threads are blocked on average can have an impact on CPU usage. Based on the above callee stacks, I'd suggest switching to an async alternate to `BlockingCollection`, perhaps [`Channels` APIs](https://devblogs.microsoft.com/dotnet/an-introduction-to-system-threading-channels/).

I see that the `ThreadPool Completed Work Item Count (Count / 1 sec)` counter shows on average more completed work items per sec on Linux, and larger spikes in the `ThreadPool Queue Length` counter on Windows, a possibility is that threads are blocking more on Windows and leading to less CPU usage.

A hopefully temporary workaround may be to increase the min worker thread count with `ThreadPool.SetMinThreads` or by other means, such as with an environment config variable ([`DOTNET_ThreadPool_ForceMinWorkerThreads`](https://github.com/steveharter/dotnet_coreclr/blob/master/Documentation/project-docs/clr-configuration-knobs.md#threadpool-configuration-knobs)) or through [project config](https://docs.microsoft.com/en-us/dotnet/core/runtime-config/threading#minimum-threads). Note that the environment variable takes values in hexadecimal. Increasing the min worker thread count may help to avoid starvation but may also increase CPU usage, see below.

I also noticed that the average throughput of queued work items to the thread pool is not very high. The thread pool may consume more CPU in lower-load scenarios (low load on the thread pool, not on the whole system). It has been improved a bit in .NET 7, and there's more room for improvement there. Setting the environment variable [`DOTNET_ThreadPool_UnfairSemaphoreSpinLimit=0`] before starting the process may help to reduce some of that CPU usage, especially when the min worker thread count is increased through config.

**Comment by kouvel** [comment]
> This change of running these long running tasks outside of the ThreadPool did not have a significant CPU load impact, however.

Is that on Linux, or Windows, or are both looking the same in CPU usage? I'm curious if the CPU usage equalized after that change.

**Comment by kouvel** [comment]
It may be useful to compare CPU profiles of Windows and Linux after that change. `dotnet-trace` may be fine if the thread pool threads are not blocking too much anymore, or otherwise `PerfView`/`perfcollect` to get on-CPU profiles.

**Comment by gregolsky** [comment]
We observe that even with `DOTNET_UnfairSemaphoreSpinLimit=0`, `.NET Timer` thread still consumes around 3% CPU on average, which is a big deal for all those very small servers for which 5% is the baseline CPU limit and going over that consumes the burst credits.

Below is a screenshot of top filtered by the main .NET process on the server. The process is an idle server - there's just some occasional heartbeats and monitoring work going. Yet the .NET timer consumes the most CPU in this scenario.

![image](https://github.com/dotnet/runtime/assets/1541805/4826a3b4-90d0-4a97-b1bc-52edc9c3d98b)

Is there anything else we can do to curb down the CPU usage on .NET Timer thread?

**Comment by gregolsky** [comment]
Is there something that the user might be doing to increase the CPU usage in this thread? 

Let me explain - we have a 2 different versions of our application, say v5 and v6. They run on the same .NET version - .NET 8. Now v6 is using order of magnitude more CPU in .NET Timer than v5. `DOTNET_ThreadPool_UnfairSemaphoreSpinLimit` is set to 0.

All it does system-wise is just a ton of constant `futex` calls. What might be different in v6 that might increase the CPU usage in `.NET Timer` thread? Where should we start looking? Please advise.
